File: 프로그래머스/lv2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.js
// 흠.. 보면 패턴이 보인다
// 1. 짝수를 2진법으로 나누면, 1의자리수는 무조건 0이기 때문에, f(x)는 x+1이 된다
  // 그 다음수는 나머지는 모두 똑같고, 1의 자리수만 0에서 1이 되기 때문이다
// 2. 홀수가 약간 골 때린다
  // x를 4로 나눈 나머지가 1인 경우의 f(x) = x+1이다
    // f(1) = 2, f(5) = 6, f(9) = 10, f(13) = 14, f(17) = 18
  // x를 4로 나눈 나머지가 3인 경우는?
    // 01부분을 10으로 나누면 된다
    // 1011 -> 1101로, 10011 -> 10101, 10111 -> 11011
    // 만약, 0이 없다면? 예로 들어서 11의 경우? 맨 앞에 0을 만들고, 바꾸면 된다
    // 11 -> 011 => '101'가 f(x)의 답이다 / 0111 -> 1011

function solution(numbers) {
  let result = []

  for (let number of numbers) {
    if (number % 2 === 0 || number % 4 === 1) result.push(number+1)
    if (number % 4 === 3) {
      if (number.toString(2).indexOf('0') === -1) { // 111같은 경우면 0111 => 1011로
        let tmp = '0'+number.toString(2)
        let tmp2 = '10'+ tmp.substr(2)
        result.push(parseInt(tmp2,2))
      } else { // 10011 -> 10101로 바꿔야 함
        let tmp = number.toString(2)
        for (let i=tmp.length-1; i>=0; i--) {
          if (tmp[i-1]+tmp[i] === '01') {
            tmp = tmp.substr(0,i-1)+'10'+tmp.substr(i+1)
            break
          }
        }
        result.push(parseInt(tmp,2))
      }
    }
  }
  return result
}

console.log(solution([2,7]))

File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.py
# 비트끼리의 비교를 어떻게 하면 효율적으로 할 수 있을까
    # 이중 포문은 에바인거 같은데
    # 이 부분은 규칙찾아야 할듯
# 흠;; 1011 vs 10101 이런 경우도 포함시켜야 하나? 01011이랑 같으니까. 2개차이남
    # -> yes


# 1트 - 실패(규칙 잘못 찾음)
# def find_bit(n):
    
#     # step1: 비트로 변환해서 쭉 돌림 -> ㄴㄴ 그냥 규칙찾음
#     # n_bit = format(n,'b') # 'o'는 8, 'x'는 16진수
#     if n % 8 == 7: return n+4*((n//8)+1)
#     if n % 8 == 3: return n+2
#     return n+1

# def solution(numbers):
#     result = []
#     for number in numbers:
#         result.append(find_bit(number))
#     return result

# print(solution([13,14,15]))
# print(solution([343])) # [347]이 나와야하는데 엉뚱한 수가 나옴 -> 규칙아예 잘못찾음;;

# 2트
# 블로그 참고;
    # 1) 짝수의 경우
    # 만약, 4라면 이진수로 100이다. 4보다 크면서 2개 이하로 다른 수를 찾으면 101
    # -> 즉, 가장 뒤에 있는 0을 1로 바꾸면 됨 -> +1

    # 2) 홀수의 경우
    # 만약, 7이라면 이진수로 0111 (바꿀때 편의를 위해 앞에 0을 붙인다)
    # 먼저 짝수의 경우처럼 가장 뒤에 있는 0의 인덱스(idx)를 찾아 1로 바꾼다 -> 1111,
    # 그런 다음 idx+1 의 인덱스 값을 0으로 바꾼다 -> 1011이 되고, 이게 바로 답.
def find_bit(num):
    if num % 2 == 0:
        return num+1
    else:
        to_bit = list('0'+format(num,'b'))
        idx = ''.join(to_bit).rfind('0') # rfind 매우 유용!!
        to_bit[idx] = '1'
        to_bit[idx+1] = '0' # 01부분을 10부분으로 바꿔주기ㅇㅇ
        return int(''.join(to_bit),2)

def solution(numbers):
    result = []
    for number in numbers:
        result.append(find_bit(number))

    return result

print(solution([13,14,15])) # 14,15,23
print(solution([343])) # [347]이 나와야하는데 엉뚱한 수가 나옴 -> 규칙아예 잘못찾음;;